Declare and override autocompleters via filter

brandonpayton opened the pull request:
## Description
This is a PR for declaring and overriding autocompleters during block registration.

## How Has This Been Tested?
I have manually tested with an ad hoc plugin, have added unit tests, and will be adding more.

## Types of changes

* Autocompleters have a `type` property so they can be identified.
* Separate option retrieval from rendering option labels and keywords, enabling us to override retrieval, labeling, and keywords separately.
* Make completers more declarative by removing their `onSelect` function and replacing it with `getOptionCompletion` that returns an action/value pair. Currently supported actions are `insert-at-caret` and `replace-block`. This detangles completers from runtime dependencies on things such as the `onReplace` function. Now, the `Autocompleter` component uses the completion action to decide how to handle the result.
* Ignore the concept of option value and simply pass the selected item to a completer's `getOptionCompletion` function. This isn't a stance against the concept of value, but it seemed like unnecessary indirection in the current implementation.
* Export a set of default autocompleters from `blocks/autocompleters`.
* Update `RichText` to include autocompletion support, including a new `autocompleters` prop that can be used to specify completers in place of the default.
* Add a `blocks/hooks/default-autocompleters` module to provide a default list of completers for blocks that don't explicitly provide their own.
* The `blocks.Autocomplete.completers` filter is applied when the `Autocomplete` component is first focused after receiving a new list of completers.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation

brandonpayton committed:
Support overriding autocomplete with filters

brandonpayton commented:
If others agree this approach is reasonable, I'd also like to add a filter for transforming the default list of completers. For my use case, I'd want to add a completer for cross-posting to different P2s (e.g., `+otherp2`).

gziolo commented on components/autocomplete/index.js:
```diff
@@ -231,13 +232,36 @@ export class Autocomplete extends Component {
+
+		// Filter the options getter itself because we only want to invoke one,
+		// whether it be default or custom.
+		const optionsGetter = applyFilters( `autocomplete.${ type }.options.getter`, completer.getOptions );
```
We need to standardize the way we name filters to have one pattern that is easy to follow. So far we were prefixing the hooks name with the package name followed by the component's name. So, in this case it would be something like: `components.Autocomplete`. Another thing here is that we can combine those 3 filters into one as follows and call it once only when the component gets initialized:

```js
this.props.completers = applyFilters( `components.Autocomplete.completers', this.props.defaultCompleters );
```

gziolo commented:
@brandonpayton, you can totally go for it. It's a valid use case and we should support a wide range of extensibility options. My only concern that your initial proposal it too granular and it would make it harder to discover for plugin developers. Another concern I have is that ideally all hooks should be applied only once when a component gets mounted. If we want to allow to dynamically add/remove/update hooks we can do it using event listeners provided by the `hooks` package. See the example Higher-Order component which is meant to hide implementation details of this behavior in a completely different context: https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-filters/index.js#L24. I would be more than happy to help to implement a similar wrapper component that would be tailored to solve the issue we have here. Let me know if you have more questions. I'm happy to discuss all the ideas I shared.

brandonpayton commented:
Thanks for the review, encouragement, and guidance, @gziolo! I think you're right about the granularity, especially for a first pass. I'll update with a single filter that runs after the component is mounted.

gziolo commented:
I noticed that @noisysocks is working on some refactoring related to Autocompleters in #4769. Cross-linking for better visibility :)

brandonpayton commented:
@noisysocks and I spoke briefly yesterday. Autocompleting reusable blocks means the completer needs access to the latest list.

I spent some time thinking today, and here's what I'm seeing.

* The `getOptions` method should simply return a promise for raw options data.
* Rendering, keywords, and value should be provided by separate functions that take an option argument. They are different concerns, and this will allow each to be overridden separately.
* Add a `components.Autocomplete.completers` filter.
* Consider adding a filter for each completer type after the above filter is applied (e.g., `components.Autocomplete.completers.blocks`).
* The default blocks completer should provide a working `getOptions` that hits the REST API.
* Via a filter mentioned above, the editor can override the blocks completer `getOptions` with one that uses the editor store.

Here's an example of what I'm imagining for the blocks completer.
```javascript
{
	completes: 'blocks',
	className: 'blocks-autocompleters__block',
	triggerPrefix: '/',
	getOptions,
	getOptionKeywords,
	renderOption,
	allowContext,
	getOptionValue,
	onSelect,
}
```

If you have time, let's talk more on Monday.

brandonpayton commented:
> * The default blocks completer should provide a working `getOptions` that hits the REST API.
> * Via a filter mentioned above, the editor can override the blocks completer `getOptions` with one that uses the editor store.

I may be wrong about this if the only expected context for a blocks completer is the editor. I've been assuming the editor only applies when editing a post, but if it also will be used in things like customization, then it probably doesn't make sense to provide a default `getOptions` that doesn't use the editor store.

I think the main thing for this PR is that option retrieval should be separated from rendering, so each can be overridden separately.

gziolo commented:
@brandonpayton, I'm looking forward to seeing this PR updated. I need to go through the code to better understand how this would work. I don't know that well the internal implementation of autocompleters :)

brandonpayton commented:
@gziolo I'm eager to work on this today but have to take care of something first. It might have to wait until tomorrow. Thanks for your comments and interest!

brandonpayton commented:
I updated the PR to...

* Separate option retrieval from rendering.
* Fix existing tests and add a couple more.
* Remove the autocomplete-related filters because I want to explore overriding autocompletion at the block registration level first.

I am working on...

* Declare autocompleters with block registration.
* Move to a single `Autocomplete` instance at the top level, likely wrapping `WritingFlow`.
  * We can look up the current block and its autocompleters based on the `input` event target.

brandonpayton commented:
I ran into serious a11y questions with moving to a single top-level Autocomplete instance and am keeping multiple instances for now.

I updated Autocomplete to look up completers based on its containing block type but still need to update the tests. I also changed completers to simply return a result for a selected option and made it the responsibility of Autocomplete to insert text or replace the current block. This is the general idea, but it probably needs some cleanup.

I'd still like a way for autocompletion to automatically apply for blocks with defined autocompleters, but I'm not sure how to do that in an accessible way since autocomplete-related aria attributes and UI should be directly associated with each input field.

brandonpayton commented:
Currently, `<Autocomplete>` looks up autocompleters by finding the containing block's DOM node, getting it's `data-type`, and looking up the block type settings. This is too tangled.

I'm thinking autocompleters should be added to the context with `BlockEdit` if they exist, possibly under the `editor` key.

brandonpayton commented:
> I'm thinking autocompleters should be added to the context with `BlockEdit` if they exist for a given block type. Then `<Autocomplete>` can use that context.

I'm thinking Autocomplete should remain a simple props-driven `@wordpress/components` component and that we should create something like a `ContextualAutocomplete` component to live under `@wordpress/blocks` and receive autocompleters from context added by BlockEdit (which will know the autocompleters registered for a block).

gziolo commented on blocks/autocompleters/index.js:
```diff
@@ -51,55 +63,54 @@ import BlockIcon from '../block-icon';
+export function blockAutocompleter() {
+	let promisedOptions;
+
+	const getOptions = () => promisedOptions || ( promisedOptions = Promise.resolve(
```
If you want to make sure this is called only once you can use `lodash.once`
```js
const getOptions = once( () => Promise.resolve... );
```

brandonpayton replied:
Thanks for the tip! That's cleaner. 
I'll make the update.

gziolo commented on blocks/autocompleters/index.js:
```diff
@@ -51,55 +63,54 @@ import BlockIcon from '../block-icon';
-	const onSelect = ( blockName ) => {
-		onReplace( createBlock( blockName ) );
-	};
+	const getOptionResult = blockData => createBlock( blockData.name );
```
I don't know this code so just wanted to double check why to remove `onReplace` here.

brandonpayton replied:
It started with a dependency issue when moving completers to block registration. We didn't yet have an `onReplace`.

Then I thought completers should be more declarative anyway. The presence of an `onSomething` name made me think about that. Why should a completer know that it is replacing something in the editor? The idea is that the editor has a better idea of what it wants to do with the result of a completer selection.

gziolo replied:
I think it ensures we remove the existing paragraph block.

brandonpayton replied:
The `Autocomplete` component now calls `onReplace` instead when a block is returned.
https://github.com/WordPress/gutenberg/pull/4609/commits/ae62091c3d822703c18fd21811dcc06696023f8a#diff-8d7b3f0b32214bb1340578e3be07ed96R176

The idea is that a completer should not be involved in making actual changes to the content.

gziolo commented on blocks/autocompleters/index.js:
```diff
@@ -108,34 +119,31 @@ export function blockAutocompleter( { onReplace } ) {
+		<span key="name" className="blocks-autocompleters__user-name">{ user.name }</span>,
+		<span key="slug" className="blocks-autocompleters__user-slug">{ user.slug }</span>,
+	];
+
```
I like this part, it makes it nicely composable out of smaller pieces.

gziolo commented on components/autocomplete/index.js:
```diff
@@ -162,14 +163,18 @@ export class Autocomplete extends Component {
+		if ( getOptionResult ) {
+			const replacement = getOptionResult( option.value, range, query );
+
+			if ( replacement && typeof replacement === 'object' && getBlockType( replacement.name ) ) {
```
Is there maybe an existing API method which validated if the input is a block? There is a similar logic in here:
https://github.com/WordPress/gutenberg/blob/master/blocks/block-edit/index.js#L22

We might want to extract method and use it in both places.

brandonpayton replied:
Agreed. I'll look at that.

brandonpayton replied:
I took a look and am wondering whether it is worth it to a function for the check. The only thing in common is that both make sure a truthy value is returned from `getBlockType`. Defining an `isBlockType` function in `blocks/api/registration` may allow us to be more direct though.

Currently, I simply moved the conditions from `components/Autocomplete` to `blocks/contextual-autocomplete` in the form of the `isReplacement` prop.

gziolo commented on blocks/autocompleters/test/index.js:
```diff
@@ -40,40 +46,55 @@ describe( 'blockAutocompleter', () => {
-			] );
+		const completer = blockAutocompleter( { } );
+		const renderedLabels = getBlockTypes().map(
+			blockType => completer.renderOptionLabel( blockType )
```
It looks like you can perform all assertions inside this loop.

brandonpayton replied:
Updated.

gziolo commented on blocks/autocompleters/test/index.js:
```diff
@@ -40,40 +46,55 @@ describe( 'blockAutocompleter', () => {
+		renderedLabels.forEach( label => {
+			// Only verify that a populated label is returned.
+			// It is likely to be fragile to assert that the contents are renderable by @wordpress/element.
+			if ( Array.isArray( label ) ) {
```
Or can we make it more explicit what is what? At the moment it might work, but it's hard to decipher what we are checking here.

brandonpayton replied:
Updated.

gziolo commented on blocks/library/paragraph/index.js:
```diff
@@ -140,10 +140,7 @@ class ParagraphBlock extends Component {
-					blockAutocompleter( { onReplace } ),
-					userAutocompleter(),
-				] }>
+				<Autocomplete getBlockType={ getBlockType } onReplace={ onReplace }>
```
We might want to iterate on `getBlockType` passed as prop. Maybe it should be a part of the autocompleter object. Not sure.

brandonpayton replied:
This will go away when I make the update for BlockEdit to provide autocompleters via context. There will be no need to find the containing block to provide autocompletion context because that will be provided naturally via React context.

gziolo commented on blocks/library/paragraph/index.js:
```diff
@@ -254,6 +251,11 @@ export const settings = {
 		],
 	},
 
+	autocompleters: [
```
It probably makes the most sense to have it here 👍 Let's see how it goes :)

brandonpayton replied:
You might want different autocompleters per block. A concrete example is that a block autocompleter makes sense in a paragraph but not in the caption field of a cover image. If you offer block completion there, it is more likely an opportunity for surprising replacement than a feature IMO.

youknowriad replied:
What If you want different autocompleters per RichText? is this a valid use case? could this be a prop of `RichText` instead?

brandonpayton replied:
I'm not sure, but it _seems_ reasonable to want different autocompleters per RichText. I'm not sure what overriding should look like in that case though.

I'm wondering whether this PR is trying to do too much and, in retrospect, would like to have created an autocompletion extensibility issue for discussion and smaller PRs. But if we can adjust in the future, maybe it would be good to continue as-is and get more feedback once merged.

youknowriad replied:
Sounds good to me @brandonpayton

gziolo commented on components/autocomplete/index.js:
```diff
@@ -231,13 +236,36 @@ export class Autocomplete extends Component {
+			const blockNode = inputNode.closest( '.editor-block-list__block' );
+			const blockName = get( blockNode, 'dataset.type', '' );
+			completers = blockName ?
+				get( getBlockType( blockName ), 'autocompleters', [] ) :
```
Yes, we might want to inject this, too. This is really interesting :)

gziolo commented:
I like where it is heading. It makes autocompleters more general. I'm wondering if we could use render-props pattern. If you don't know the concept I recommend this article: https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce or downshift library: https://github.com/paypal/downshift. We used it also with the `<Dropdown />` component: https://github.com/WordPress/gutenberg/tree/master/components/dropdown in Gutenberg. It might make sense to follow the same pattern in this refactoring. I didn't investigate how this could be done, but this is very powerful when helping decouple what autocompleter neeeds and how this should be represented in the UI. I think it's pretty close to this after you applied all the recent changes.

brandonpayton commented:
I updated this PR with the following...

* Updated `BlockEdit` to add a block's `autocompleters` to the context.
* Added a `blocks/contextual-autocomplete` component based on `components/autocomplete` to leverage the `autocompleters` context.
* Updated `blocks/library/paragraph` to use `contextual-autocomplete` instead of the props-driven version.
* Replaced the `getBlockType` dependency of `components/autocomplete` with a new `isReplacement` prop to allow the component to ask whether it has a replacement to pass to `onReplace`.

I took a quick look at the render-props pattern but need to consider more where it could help us here. I believe `components/autocomplete` does something similar for accessibility reasons but uses the `children` prop:
https://github.com/brandonpayton/gutenberg/blob/f0f3c5bd5cd18c3a03212fb406e78f5c77d3d9f9/components/autocomplete/index.js#L486

gziolo commented on blocks/block-edit/index.js:
```diff
@@ -18,35 +19,47 @@ import {
+export class BlockEdit extends Component {
+	getChildContext() {
+		const blockType = getBlockType( this.props.name );
+		return { autocompleters: blockType.autocompleters };
```
I see what you did. This is also a way to go.

brandonpayton replied:
It seems like I missed something in our conversation yesterday, because I thought this is what we talked about. :)  Perhaps I missed your intent with the render-props pattern.

If we continue to use context, I'm thinking it would be cleaner and more generally useful to add the block type name or settings to the context rather than autocompleters specifically. Once you have the block type, it is straightforward to look up what you need.

gziolo replied:
Context are fine, see you what I did today when testing React 16.3 and new Context API: https://github.com/WordPress/gutenberg/pull/5029.

I didn't mean we shouldn't do it. I just hoped we can move `<Autocomplete />` outside of `edit` component. It seems like it would be too tricky to achieve. Let's continue with the approach we have at the moment 👍

brandonpayton replied:
Nice use of context IMO. I love the new API.

gziolo commented on blocks/library/paragraph/index.js:
```diff
@@ -160,10 +160,7 @@ class ParagraphBlock extends Component {
-				] }>
+				<ContextualAutocomplete onReplace={ onReplace }>
 					{ ( { isExpanded, listBoxId, activeId } ) => (
 						<RichText
```
Can we wrap `RichText` with `ContextualAutocomplete`? We have now access to autocompleters through context so this should work :)

gziolo replied:
Definitely something to think about in the future. Let’s see how many RichText components will we wrap. It’d be more useful for plugin developers as they would easily update core components, but sometimes there might be a few RichText components bundled together, like in case of `Quote`.

brandonpayton replied:
> Can we wrap RichText with ContextualAutocomplete? We have now access to autocompleters through context so this should work :)

I was thinking about that too because I wanted block authors to get autocompletion without manually wrapping their inputs with an autocomplete component. I wonder if we should just define `ContextualAutocomplete` within `RichText` and use it to directly wrap `<TinyMCE>`.

That brings to mind another lingering question:
Should we support declaring default autocompleters?

I'm thinking the user autocompleter would be desired everywhere RichText is used, while the block autocompleter could be odd or surprising.

gziolo replied:
Yes, it raises more questions. I think we should continue as it is, but revisit later when we have more blocks using this feature.

brandonpayton replied:
I didn't have time to spend on this today, but for a good first implementation of extensible autocompletion, I'm thinking we should do the following and then polish:

* Move `<ContextualAutocomplete>` out of the paragraph block and into `RichEdit`, wrapping `TinyMCE`. I believe this would give us fairly good autocompletion support out of the box.
* Provide a set of default autocompleters that are assigned to each registered block without declared autocompleters. It would probably be just the user completer for now.
  * We could support filtering the default set of autocompleters later.
  * I don't love actually applying the defaults to the block settings, but while we aren't providing a default autocompleters filter, it would allow a `blocks.registerBlockType` filter to override autocompleters for any block type.

You mentioned
> I think we should continue as it is
but I'm not 100% whether that means you'd rather not do the items above. What do you think?

brandonpayton commented:
@gziolo I updated the PR to move `ContextualAutocomplete` into `RichText` and added a `default-autocompleters` hook to add default autocompleters during block registration. It broke some snapshot tests and also broke some tests involving registered blocks because the expected block settings don't include autocompleters. I can look at fixing those tomorrow but wanted to make these changes available to you today.

brandonpayton commented:
Wow, that's embarrassing. I forgot to push the changes I mentioned yesterday. They're pushed now.

gziolo commented:
@aduth or @youknowriad, can you double check if this is good to proceed as is? I'm busy with other work at the moment.

brandonpayton commented on blocks/autocompleters/index.js:
```diff
@@ -51,55 +63,52 @@ import BlockIcon from '../block-icon';
 		triggerPrefix: '/',
 		getOptions,
+		getOptionKeywords,
+		renderOptionLabel,
```
The reason is that they are separate concerns. Otherwise, if you want to override label rendering, you would also have to take responsibility for option retrieval and keywords. This way, you can override retrieval, rendering, and keywords separately.

youknowriad replied:
I think I have a small prefernece for the old API here, it looked simpler and easier to understand. 
Is it really a common use-case to replace just one part of the autocompleter? My thinking is that the autocompleters are simple to write and in general we would provide new ones instead of replacing a property? (even if there's some duplication)

brandonpayton replied:
> I think I have a small prefernece for the old API here, it looked simpler and easier to understand.

I'm sympathetic to this but am thinking it is better to separate concerns. Currently, if a developer wants to simply tweak an option label, they also have to take responsibility for retrieving options and providing option keywords. It requires that they know and understand what the default behavior so they can provide an default-like experience that only overrides the label. If the default changes and they do not notice, the custom completer will unintentionally differ from the default.

Since `getOptions` requires the implementor to think about providing a rendered `label` and `keywords` in addition to option retrieval, I'm not sure it is more complex in practice than having dedicated functions for label and keywords. They are still things that must be considered, but with the updated interface, they can be overridden and tested separately.

I'm not sure I've done a great job at selling this, so any more thoughts here?

youknowriad replied:
I understand the concerns but I guess the answer to this question would give us the best option here.

> Is it really a common use-case to replace just one part of the autocompleter?

I personally don't think it's common use-case but I may be wrong. And it's not a blocker for me anyway in this PR :)

brandonpayton commented on blocks/autocompleters/index.js:
```diff
@@ -51,55 +63,52 @@ import BlockIcon from '../block-icon';
+	const getOptionResult = blockData => createBlock( blockData.name );
 
 	return {
+		type: 'block',
```
That's a good question. My intent was to mark completers with what they complete, but it might be a good idea to provide an ID instead. I could see cases where someone wants to override a core autocompleter but not one added by a plugin.

What do you think?

noisysocks replied:
Yeah, that's a good point. Something that identifies this specific autocompleter, e.g. `name: 'core-autocompleter/block'` works for me.

brandonpayton commented on blocks/block-edit/index.js:
```diff
@@ -18,35 +19,47 @@ import {
+	autocompleters: noop,
+};
+
 export default withFilters( 'blocks.BlockEdit' )( BlockEdit );
```
The purpose of adding to the context is to allow using `BlockContextualAutocomplete` within blocks (or their child components like `RichText`) without having to look up and explicitly pass the current block's autocompleters.

The idea for a first iteration here is to allow overriding autocompleters via a `blocks.registerBlockType` filter. I initially added autocompletion-specific filters, but after conversation with @gziolo, thought using an existing hook was a better first step.

brandonpayton commented on blocks/block-edit/index.js:
```diff
@@ -18,35 +19,47 @@ import {
+
+		// `edit` and `save` are functions or components describing the markup
+		// with which a block is displayed. If `blockType` is valid, assign
+		// them preferencially as the render value for the block.
```
Fixed!

brandonpayton commented on blocks/block-contextual-autocomplete/index.js:
```diff
@@ -0,0 +1,21 @@
+
+BlockContextualAutocomplete.defaultProps = {
+	isReplacement( completion ) {
+		return completion && typeof completion === 'object' && getBlockType( completion.name );
```
My reasoning is `Autocomplete` is currently a base component that doesn't know about blocks and wouldn't have any block-related imports.

As a side note, I'd rather provide `onReplace` via context here as well so users of `BlockContextualAutocomplete` don't have to pass it along. The default would likely be `noop`.

youknowriad replied:
Why the new React context API, this component might not be necessary anymore, did you try it?

brandonpayton replied:
I'd _really_ rather use the new API than introduce this new component, but I'm not sure when 16.3 will be released so we can use it. I will try to think of a good way to set this aside until then.

brandonpayton commented on components/autocomplete/index.js:
```diff
@@ -149,7 +150,7 @@ export class Autocomplete extends Component {
 	}
 
-	replace( range, replacement ) {
+	completeText( range, replacement ) {
```
Ah, I was not aware of that. Thank you.

brandonpayton replied:
I renamed this to `insertCompletion`. It's technically a replacement, but IMO the name is clearer than a generic `replace`.

youknowriad commented on blocks/autocompleters/index.js:
```diff
@@ -51,55 +63,52 @@ import BlockIcon from '../block-icon';
 
-	const onSelect = ( blockName ) => {
-		onReplace( createBlock( blockName ) );
-	};
```
Removing this creates a need to check that the result is a block outside of the autocompleter. Imagine an autocompleter that returns a block but we don't way to replace, we want to append something.

Three options:

 - keep onSelect
 - add and `action` property that could have things like `replace`, `append`, `insert-at-caret` values...

brandonpayton replied:
Today, the block completer has a runtime dependency on a block's `onReplace`. The paragraph block provides `onReplace` when it creates the block completer, but it's less clear to me how to satisfy such dependencies while allowing completers to be overridden during block registration.

I started thinking that maybe completers should simply provide completions without imperative logic regarding what to do with a completion. It's better to keep things simpler and less entangled, but I see a couple of problems with my thinking:

1. You make a good point that we may want to do other things with completions, like append. I'm somewhat attracted to the option of tagging completers with actions like 'insert-at-caret', 'replace', and 'append' but am thinking it's probably a mistake to constrain completers to declarative options.
2. For the block completer to include reusable blocks (see #3791), I believe we need to satisfy a runtime dependency on the editor state, so removing `onSelect` does not directly address the question of completer dependencies. My thinking for this was to add a `blocks.registerBlockType` filter to override the block completer's `getOptions` with a version based on the editor store. This is hacky IMO.

@youknowriad Do you have any thoughts on how we could provide runtime dependencies to completers in a cleaner, more generic way?

youknowriad replied:
> My thinking for this was to add a blocks.registerBlockType filter to override the block completer's getOptions with a version based on the editor store. This is hacky IMO.

Actually, I don't think it's hacky, it means that if you use the blocks module only, you get the generic blocks, if you use the editor blocks you get the generic  + reustable blocks. I think this goes in concert with the need to move the reusable blocks definition to the editor module as well. cc @noisysocks 

The runtime dependency we need to satisfy here `onReplace` is different though because it's not a global dependency, it's contextual to the block. and I'm fine with tagging completers to solve this => this means the generic autocompleter behavior is block-aware but it's fine I think.

brandonpayton commented:
@youknowriad I updated this PR based on our discussion and did the following:

* Updated the Autocomplete component to respond to `insert-at-caret` and `replace-block` completion actions rather than simply inferring replacement from the fact that the completion is a block.
* Fixed existing unit tests
* Added unit tests
* Implemented a React-like `createContext` function while we are waiting on the actual context API in 16.3. I did this for fun but thought it was cool to actually be able to write a context provider and consumer in this style. Want to leave it or remove it?
* Created an AutocompletionContext. BlockEdit uses its Provider, and BlockContextualAutocomplete uses its Consumer. I thought it best to keep BlockContextualAutocomplete to represent the fact that autocompletion in blocks is contextual by default.

I need to update docs to show declaring and overriding autocompleters.

There is an e2e test failure due to another element overlapping a click target. I don't have time to fix it tonight but plan to look at it in the morning.

Do you see anything else we need to do here?

brandonpayton commented:
The e2e test failure seems to have been a fluke. I've seen some inconsistency in e2e tests locally as well.

aduth commented on components/autocomplete/index.js:
```diff
@@ -162,15 +163,19 @@ export class Autocomplete extends Component {
+		if ( getOptionCompletion ) {
+			const { action, value } = getOptionCompletion( option.value, range, query );
+
+			if ( 'replace-block' === action ) {
```
`components` are UI primitives not specific to any screen. There should be no understanding of what a block is at this point.

brandonpayton replied:
Good point. Changing this to 'replace' would skirt that. Is that enough? (It was @youknowriad's original suggestion anyway, but I momentarily forgot this should not be block-aware)

aduth commented on blocks/context/index.js:
```diff
@@ -0,0 +1,53 @@
+import { Component, Fragment } from '@wordpress/element';
+
+/**
+ * This is a stopgap until React 16.3 is released with the new context API.
```
Can we just pull in the React alpha? I'd personally be fine with it. (We ran React 16.0.0 alpha for many months)

brandonpayton replied:
That would be much better. I'll do that.

brandonpayton replied:
Well, I'll do that if our approach to autocompletion leaves us with contextually defined autocompleters.

gziolo replied:
I updated https://github.com/WordPress/gutenberg/pull/5029 which contains all required changes to bring in React 16.3. You can cherry pick the initial commit from this PR or merge the full PR as it is. I tried to make it ready to merge with all the required docs updates and so on. Just letting you know. There is also https://www.npmjs.com/package/create-react-context polyfill that we could use temporarily.

aduth commented on blocks/block-edit/index.js:
```diff
@@ -26,6 +27,8 @@ export function BlockEdit( props ) {
 		return null;
 	}
 
+	const { autocompleters = [] } = blockType;
```
For blocks which don't define their own `autocompleters`, this assign a new reference value every single render, which I assume will then trickle down to `AutocompletionContext.Provider` as a new value (and update its context accordingly).

Could we...

- Gracefully handle an undefined `autocompleters` in the provider component?
- Move the default value up to the module scope so it's a shared value.

aduth commented on blocks/block-edit/index.js:
```diff
@@ -34,18 +37,20 @@ export function BlockEdit( props ) {
-			focus={ props.isSelected ? {} : false }
-			setFocus={ noop }
-		/>
+		<AutocompletionContext.Provider value={ { autocompleters } }>
```
Similar note to above, we're passing a new object reference every single render, which will break any attempt at purity reconciliation if it exists (and based on my limited understanding of React context providers, likely trigger a re-render to all consumers). Can we not have this prop shaped as an object? Simpler type for props is better.

brandonpayton replied:
I'll test this. My believe was that it would only affect the provider's children. I'm using the provider incorrectly if this affects anything other than descendant consumers.

aduth commented on blocks/context/index.js:
```diff
@@ -0,0 +1,53 @@
+
+		render() {
+			return (
+				<Fragment>
```
You shouldn't need a `<Fragment>` here, just:

```js
return this.props.children;
```

brandonpayton replied:
Good to know. Thanks!

aduth reviewed (commented):
I agree with the general sentiment that autocompleters need to be extensible but... I expect autocompletion as a behavior specific to RichText. It's odd to see it defined as a property of a block. We can expect many blocks will not have any RichText input.

How it existed previously as far as usage in a block seemed fine to me, so personally I view this as a step backward.

brandonpayton commented on blocks/api/test/registration.js:
```diff
@@ -82,7 +82,7 @@ describe( 'blocks', () => {
 			const block = registerBlockType( 'my-plugin/fancy-block-4', defaultBlockSettings );
 			expect( console ).not.toHaveErrored();
-			expect( block ).toEqual( {
+			expect( block ).toMatchObject( {
```
My goal in this change was to simply test block registration, but block registration hooks are already registered at this point and affect the block settings. Do you agree, and if so, do you have any suggestions about how to handle this?

One possibility I see but don't love is to remove all `blocks.registerBlockType` filters and expect unit tests for those filters to register and remove them. This may require additional exports from hook modules for the sake of testing.

brandonpayton commented:
> I agree with the general sentiment that autocompleters need to be extensible but... I expect autocompletion as a behavior specific to RichText. It's odd to see it defined as a property of a block. We can expect many blocks will not have any RichText input.
> 
> How it existed previously as far as usage in a block seemed fine to me, so personally I view this as a step backward.

Thank you for the review, @aduth.

Declaring and overriding autocompleters at the block level bothered me a bit because we may need more granularity within multi-field blocks and, as you said, not all blocks require autocompletion. But I started heading this direction because it seemed better to override once per block type than to override autocompleters for each block or RichText instance that used them. In addition, it seemed reasonable to want to override completers for particular block types, so wherever we apply filters for overriding autocompletion, we'd want to provide additional contextual information.

Do you have any ideas for a better approach or a direction to explore?

cc @mtias

brandonpayton commented:
This PR initially offered the ability to filter the completers passed to the Autocomplete component. I moved away from that because it seemed like we should not apply filters for each mounted instance, but we have another option of filtering on-demand when the focus first enters an Autocomplete instance. While the filter would be applied for each focused Autocomplete instance, it would be gradual and reasonable IMO.

I think we may eventually want to provide contextual information like block type and settings to an autocompleters filter, but that can be done later if there is need.

My updated plan is to:

* Continue including `<Autocomplete>` within `<RichText>`.
* Support an `autocompleters` prop on `<RichText>`.
* Use default `autocompleters` if none are provided.
* Apply an autocompleters filter when focus first enters an `<Autocomplete>` component.
* Continue with separated `getOptions`, `renderOptionLabel`, and `getOptionKeywords` functions.
* Continue with switching from `onSelect` to `getOptionResult` to decouple completers from the application of a completion.

Any feedback is appreciated.

gziolo commented:
> Apply an autocompleters filter when focus first enters an <Autocomplete> component.
`withFilter` HOC should take care of it

The plan looks solid.

When testing `createContext` I discovered that it isn't also supported by Enzyme (see #1509) so it might be difficult to create unit tests for the code that uses Provider + Consumer approach.

brandonpayton commented:
I added more documentation, including a section in docs/extensibility.md. The autocompleters hook could still use some JSDoc.

brandonpayton commented:
@pento and I reviewed this together and generally agreed on the approach, with a few improvements to the API. I'll make those changes before requesting another review.

brandonpayton commented:
I updated this PR according to review notes made by @pento and believe it is ready for another review.

* Rename `type` property to `name`
* Default to inserting completion. Don’t require action/value pair from `getOptionCompletion`.
* Allow filters to mutate completers
* Rename `renderOptionLabel` to `getOptionLabel` for consistency.
* Rename `getOptions` to `options` and allow it to be an array, a function that returns an array, or a function that promises an array. This makes the API easier to use while we internally continue to use a single async code path for consuming completer options.

cc @gziolo

gziolo commented on docs/extensibility.md:
````diff
@@ -153,3 +153,90 @@ _Note:_ This filter must always be run on every page load, and not in your brows
+{% ES5 %}
+```js
+// Our completer
+function abbreviationCompleter( completers ) {
````
`completers` is not used inside the function, should be removed.

gziolo commented on docs/extensibility.md:
````diff
@@ -153,3 +153,90 @@ _Note:_ This filter must always be run on every page load, and not in your brows
+{% ESNext %}
+```js
+// Our completer
+function abbreviationCompleter( completers ) {
````
`completers` is not used inside the function, should be removed.

gziolo commented on blocks/hooks/autocompleters.js:
```diff
@@ -0,0 +1,101 @@
+			}
+		}
+
+		updateFilteredCompleters( nextCompleters ) {
```
It can be also simplified to:
```js
updateFilteredCompleters( nextCompleters = defaultCompleters ) {
```
by using a default param value.

brandonpayton replied:
Thanks, I'm not sure how I missed that :).  Based on one of your other comments, I've updated the PR to provide defaults via a filter instead, so this no longer applies.

gziolo commented on blocks/hooks/autocompleters.js:
```diff
@@ -0,0 +1,101 @@
+
+			this.setState( {
+				hasStaleCompleters: false,
+				completers: applyFilters(
```
There is also `hasFilter` method which would be useful here to avoid state updates when there are no filters registered.

brandonpayton replied:
In this case, the state must always be updated after new completer props are received because the filtered completers are maintained in the state. This is done on demand when the component first contains focus after receiving new completer props.

We can still use `hasFilter` to avoid copying the completers list when there is no filter added, and I'm making that change now.

gziolo commented on blocks/hooks/autocompleters.js:
```diff
@@ -0,0 +1,101 @@
+ * @param  {Function} Autocomplete Original component.
+ * @return {Function}              Wrapped component
+ */
+export function filterAutocomplete( Autocomplete ) {
```
There is a convention to prefix all HOCs with `with` prefix. In this case, it might be better to name it `withFilteredAutocompleters`.

gziolo commented on blocks/hooks/test/autocompleters.js:
```diff
@@ -0,0 +1,151 @@
+			wrapper.find( 'input' ).simulate( 'focus' );
+			expect( completersFilter ).toHaveBeenCalledWith( expectedCompleters );
+		} finally {
+			wrapper.unmount();
```
It'd be less code to put `wrapper.unmount()` inside `afterEach` call instead of using try / finally in all tests.

brandonpayton replied:
I was actually doing it that way initially but switched away because I thought I would be able to use `shallow` for most of the tests. I was wrong. :)

I started to prefer the explicit try/finally because it is easy to accidentally declare a local `const wrapper` that shadows the outer variable, but I just noticed we have a `no-shadow` lint rule that covers this case.

I'll switch it back.

gziolo commented on blocks/hooks/autocompleters.js:
```diff
@@ -0,0 +1,101 @@
+ */
+import defaultCompleters from '../autocompleters';
+
+/**
```
As far as I understand what happens here, it seems like it would be enough to have only one filter registered for autocompleters: `components.Autocomplete.completers`.

Most of the logic that is inside this file is directly related to the Autocomplete component. Instead of adding filter on the `component.Autocomplete`, it is also possible to set default autocompleters using:
```js
const setDefaultAutocompleters = () =>  [ ...defaultCompleters ];
addFilter(
    'components.Autocomplete.completers',
    'core/autocompleters/set-default-autocompleters',
    setDefaultAutocompleters
);
```

brandonpayton replied:
I'm not sure I correctly understand your comment, but my reading is that this component wrapper may not be necessary. Is that correct?

The purpose of this module is to filter autocompleters passed to Autocomplete components and to provide default completers when none are provided. The filtering is done on-demand per Autocomplete component when focus first enters the component after receiving new completers via props.

Your comment made me realize it would be more idiomatic to provide default completers via a `components.Autocomplete.completers` filter, and I've made that update.

gziolo reviewed (commented):
I started the review. I left a few comments. I will download all code tomorrow and perform some testing.

brandonpayton commented:
I added the first version of a README to `blocks/autocompleters` and plan to follow up on review comments first thing tomorrow.

brandonpayton commented on blocks/library/button/test/__snapshots__/index.js.snap:
```diff
@@ -7,12 +7,22 @@ exports[`core/button block edit matches snapshot 1`] = `
-      contenteditable="true"
-      data-is-placeholder-visible="true"
-    />
+    <div>
```
I noticed this as well and was initially puzzled, but they are divs added by the Autocomplete component now used within RichText:

* `Autocomplete` adds a div to wrap TinyMCE, adding event listeners and a sibling popup.
* `withFocusOutside` wraps `Autocomplete` and adds a parent div because it adds event listeners.
* The autocompleters hook adds another parent div in order to add its own event listeners.

I dislike this but do not currently see a better solution. I would love to be able to add event listeners to a Fragment (mentioned [here](https://reactjs.org/docs/fragments.html#keyed-fragments) as a possible future capability) to avoid rendering the elements. I'll give it some thought. Do you have any suggestions?

gziolo commented on components/autocomplete/index.js:
```diff
@@ -3,6 +3,7 @@
  */
 import classnames from 'classnames';
 import { escapeRegExp, find, filter, map } from 'lodash';
+import 'element-closest';
```
`wp.components.Autocomplete` was exposed for some time. Is there any way to make it backward compatible for the next 2 release cycles? We should also add a deprecation message to give a warning about the changes.

brandonpayton replied:
We should be able to detect a difference in the completer objects and log a deprecation message then. Do we want to limit such messages or print for every completer object seen with the old interface?

gziolo replied:
In other cases, it was done per function/component occurrence. See #5833 for reference.

gziolo commented on blocks/autocompleters/README.md:
```diff
@@ -0,0 +1,117 @@
+Autocompleters
```
I have a feeling that this should be documented next to the original `Autocomplete` component. It is also breaking change as far as I understand. We should provide some notes how to update code when someone was using `wp.components.Autocomplete`.

brandonpayton replied:
> We should provide some notes how to update code when someone was using wp.components.Autocomplete.

Where is a good place to provide such notes?

gziolo replied:
> I plan to move the completer interface JSDoc into components/autocomplete/index.js. Sound reasonable?

Yes, this was my exactly my point. It's a general interface that should work with every custom `Autocomplete` component.

gziolo commented on blocks/autocompleters/index.js:
```diff
@@ -39,103 +45,41 @@ import BlockIcon from '../block-icon';
-		onSelect,
-	};
-}
+export default [ userAutocompleter ];
```
I'm not convinced that we should have the default setup exposed in here. We should rather set it in the hook.

brandonpayton replied:
Agreed. I think that is a better idea. I'll make a note to make the update once we decide what to do instead of the `components.Autocomplete` filter.

gziolo commented on blocks/hooks/autocompleters.js:
```diff
@@ -0,0 +1,114 @@
+	return completers;
+}
+
+addFilter( 'components.Autocomplete', 'core/autocompleters', withFilteredAutocompleters );
```
I still don't understand why we need `components.Autocomplete` filter. It is applied globally so it wraps all autocompleters. Could it be applied directly in `components/Autocomplete`? Or should we alternatively create `BlockAutocomplete` component and use it inside `RichText`?

brandonpayton replied:
I don't believe there is anything here that requires the `components.Autocomplete` filter. I am trying to remember my frame of mind when I wrote this. In [this comment](https://github.com/WordPress/gutenberg/pull/4609#issuecomment-372267251), I may have thought you were suggesting we add this functionality separately via such a filter, so it is added when blocks are in use. Now that you mention it though, I suppose block and non-block UI may exist simultaneously so that we don't want to do this for all instances.

Should autocompletion extensibility be a feature limited to `@wordpress/blocks` or also supported for `@wordpress/components`?

My mind has been set on blocks, and I have not given much thought to WordPress UI outside of blocks. I don't yet know what I think about that question, but an answer will help us know what to do instead of this unnecessary filter.

If it is a blocks-only feature, I think creating a `BlockAutocomplete` component makes sense to provide the on-focus behavior currently added via the `components.Autocomplete` filter, and we should also use a blocks-related filter name instead of the generic `components.Autocomplete.completers`.

I'm thinking that the opinion of a default completers list should also be limited to blocks, but if we want autocompleters to be extensible for `@wordpress/components` as well, I am not sure how to filter and provide defaults only within a block context because the filter could be applied in both block and non-block contexts.

gziolo replied:
I went through a similar thinking process for `MediaUpload` component in #5846. I wanted to move it to `components` module, but the feedback I received suggests to leave it inside `blocks` module together with the accompanying filer. I think it makes a lot of sense because the idea is to have a set of independent components as @youknowriad explained:
> I'm not sure we want to move components relying on filters to the components folder. I feel like the components folder is for generic React components while adding filters adds an unwanted dependency the `wp hooks`.

I would follow that advice and remove the `components.Autocomplete` filter and introduce `BlockAutocomplete` component which uses `components.Autocomplete.completers` internally.

gziolo reviewed (commented):
I left a few more comments, which circle back to my latest comments. In addition, I would like to make sure we have a good strategy for rollout. We should do our best to inform plugin developers about breaking changes introduced.

brandonpayton commented on docs/extensibility.md:
```diff
@@ -153,3 +153,86 @@ _Note:_ This filter must always be run on every page load, and not in your brows
 
 Refer to [the plugins module documentation](../plugins/) for more information.
+
+## Overriding autocompleters
```
OK, I will move them in an `autocompletion.md` doc. Does that work for you?

gziolo replied:
It all depends if you decide to have it as block specific or component specific :)

Here is the document for [Extending Blocks](https://github.com/WordPress/gutenberg/blob/master/docs/extensibility/extending-blocks.md).

It might be also its own `Extending Components` document. Autocompletion might be too granular.

gziolo commented on blocks/library/paragraph/test/__snapshots__/index.js.snap:
```diff
@@ -2,30 +2,32 @@
-        />
-        <p
-          class="blocks-rich-text__tinymce wp-block-paragraph"
+  <div
```
So it looks like this issue with the obsolete `div` elements isn't something new. Let's leave it as it is and circle back later. We can investigate if React 16.3 has something to offer to mitigate it. https://reactjs.org/blog/2018/03/29/react-v-16-3.html - introduces `createRef` and `forwardRef` - they might be what we need in this case. We still need to expose those methods in a separate PR.

gziolo commented on components/autocomplete/completer-compat.js:
```diff
@@ -0,0 +1,62 @@
+		compatibleCompleter.className = deprecatedCompleter.className;
+	}
+	if ( 'allowNode' in deprecatedCompleter ) {
+		compatibleCompleter.allowNode = ( ...args ) => deprecatedCompleter.allowNode( ...args );
```
Could you map other props like this?
```js
const extraProps = [ 'className', 'allowNode', 'allowContext' ].
    filter( key => key in deprecatedCompleter ).
    map( key => deprecatedCompleter[ key ];

const compatibleCompleter = {
    name: generateCompleterName(),
    // ...
    ...extraProps,
};
```
It seems like those functions don't add any new logic.

brandonpayton replied:
Yep, that's more concise. I wanted to preserve completer context for the methods in that list, but `Autocomplete` already doesn't call them with completer context.

gziolo commented on blocks/autocompleters/test/block.js:
```diff
@@ -0,0 +1,93 @@
+	} );
+
+	it( 'should derive option keywords from block keywords and block title', () => {
+		expect.hasAssertions();
```
It isn't that useful in this test. In the previous one it makes a lot of sense though.

brandonpayton replied:
That's very odd. Good catch.

gziolo commented on blocks/autocompleters/index.js:
```diff
@@ -1,141 +1,7 @@
-	};
-}
+export { default as blockAutocompleter } from './block';
+export { default as userAutocompleter } from './user';
```
Nice, thanks 👍

gziolo commented on components/autocomplete/README.md:
```diff
@@ -0,0 +1,121 @@
+Autocomplete
+============
+
+TODO
```
It can be very brief - usually, one line is enough :)

brandonpayton replied:
Adding a one line description works for me. :) I was thinking I needed to document props for the `Autocomplete` component as well since the README didn't yet exist, but that doesn't have to be part of this PR.

gziolo replied:
Follow up is fine

gziolo commented on components/autocomplete/README.md:
````diff
@@ -0,0 +1,121 @@
+		<abbr title={ option.name }>{ option.visual }</abbr>
+	),
+};
+```
````
Can we also add a line where it is wired into an actual component?

brandonpayton replied:
I created a simple example wiring to `Autocomplete` wrapping an `input` element, but it didn't work. It seems like `Autocomplete` is assuming a contenteditable target, since it inserts HTML rather than simple text. I'll play with a simple contenteditable example but don't have anything yet.

gziolo commented on blocks/autocompleters/README.md:
```diff
@@ -0,0 +1,117 @@
+Autocompleters
+==============
+
+Autocompleters enable us to offer users options for completing text input. For example, Gutenberg includes a user autocompleter that provides a list of user names and completes a selection with a user mention like `@mary`.
```
Can we add a reference to `@wordpress/components`? It contains the same content and will be difficult to keep in sync otherwise.

brandonpayton replied:
I actually meant to remove that entirely after moving the content but missed doing so last night.

brandonpayton replied:
Leaving a README with a reference to `@wordpress/components` seems good. Is there a better way than direct GitHub links to link to component READMEs?

gziolo replied:
Unfortunately not yet :(

gziolo commented on blocks/autocomplete/index.js:
```diff
@@ -0,0 +1,103 @@
+			const hasFocus = this.parentNode.contains( document.activeElement );
+
+			// This may trigger another render but only when the component has focus.
+			if ( hasFocus && this.hasStaleCompleters() ) {
```
Can you explain why we need this additional check in here? It seems like having `onFocus` handler should be enough.

brandonpayton replied:
It is technically possible for the `completers` prop to change while the component contains the focus. This check exists for that case, since `onFocus` will already have been called. I don't think that is likely to occur, but I'd rather just cover the case. Sound OK?

gziolo replied:
I hope it’s good 😃It makes sense, might make sense to leave an inline comment.

brandonpayton replied:
It's mostly about not assuming when props will be updated. Sure, I can add a comment. :)

gziolo reviewed (commented):
I left another round of comments :)

gziolo commented:
I noticed that this PR introduces regression for the following use cases:
* Image and Gallery - it is not possible to update the `caption` field.
* Quote and Pullquote - it is not possible to type in the `cite` part.

brandonpayton commented on docs/extensibility/autocompletion.md:
````diff
@@ -0,0 +1,85 @@
+```js
+// Our completer
+var acronymCompleter = {
+	name: 'acronyms',
````
I've struggled with whether to make name singular or plural. It feels weird to call it "acronym" because it is not an acronym, it completes acronyms. I ended up using plural for the user and block completers and wanted to be consistent here. I'm fine changing it to singular everywhere though. Any thoughts?

gziolo replied:
Let’s keep it consistent. No need to update  👍

brandonpayton commented:
I fixed the regressions and made updates in response to review feedback.

I'd like to add a couple more automated tests, but this is ready for another round of manual testing.

mkaz commented:
Looking good, I've tested it out and auto complete is working in lists, captions, quotes, heading and I assume other RichText components. Nice work!

gziolo commented on components/autocomplete/completer-compat.js:
```diff
@@ -0,0 +1,56 @@
+		link: 'https://github.com/WordPress/gutenberg/blob/master/components/autocomplete/README.md',
+	} );
+
+	const optionalProperties = [ 'className', 'allowNode', 'allowContext' ]
```
It needs to be converted to object, my fault :)

brandonpayton replied:
I did the same thing. It's funny how our minds can just fill in the gaps sometimes. Thanks for fixing it.

gziolo commented:
I'm seeing the following error once per 20ish focus changes:

> Uncaught TypeError: Cannot read property 'top' of undefined
>     at RichText.getFocusPosition (index.js:437)
>     at RichText.onNodeChange (index.js:635)
>     at c.i [as fire] (tinymce.min.a3a2e6f1.js:8)
>     at w.fire (tinymce.min.a3a2e6f1.js:8)
>     at Object.nodeChanged (tinymce.min.a3a2e6f1.js:11)
>     at w.nodeChanged (tinymce.min.a3a2e6f1.js:12)
>     at w.<anonymous> (tinymce.min.a3a2e6f1.js:11)
>     at c.i [as fire] (tinymce.min.a3a2e6f1.js:8)
>     at w.fire (tinymce.min.a3a2e6f1.js:8)
>     at x.g (tinymce.min.a3a2e6f1.js:8)

![screen shot 2018-04-05 at 11 55 11](https://user-images.githubusercontent.com/699132/38359303-3ca5c9c8-38c8-11e8-9b7f-7bbc83565a9c.png)

![screen shot 2018-04-05 at 11 54 30](https://user-images.githubusercontent.com/699132/38359262-1ff2d12c-38c8-11e8-815f-8eef37bcb3e8.png)

It doesn't break anything, but it would be nice to make sure it isn't triggered.

gziolo commented:
Tried with the legacy user completer and it shows a very nice message:

![screen shot 2018-04-05 at 12 02 40](https://user-images.githubusercontent.com/699132/38359762-535ed00a-38c9-11e8-9e55-cc8a70d3d638.png)

Nice job in this area 👍

gziolo committed:
Components: Small tweak to the Autocomplete component

gziolo reviewed (approved):
We went through so many iterations and it looks really solid. I tested it quite long today and didn't find any issues. It is a really great improvement 👍 

I made sure that deprecation message is printed to the console when using old autocomplete intereface.

I also tested acronym completer example applied with hook. It works like a charm. It even gets applied just after I install it using JS console.

Great work. There are some random errors that might be not related to this PR, but it would be nice if you would try to reproduce them.

Time to merge 🎉

brandonpayton commented:
> Great work. There are some random errors that might be not related to this PR, but it would be nice if you would try to reproduce them.

Sure, I'll play with this to see what I can find.

Thanks for fixing the silly mistake with array spread instead of object spread in the completer compat module. Also thank you for all the time you gave reviewing and testing this. 🙇
